File: ColdWarPrototype/GameWorld/CwpProvince.vb
'    OpenSGSGLibrary is an open-source library for Grand Strategy Games
'    Copyright (C) 2019  Torben Kneesch
'
'    This program is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    This program is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports OpenGSGLibrary.WorldData

Namespace GameWorld

    ''' <summary>
    ''' Province class for cold war game. Adds specialised properties to base province class.
    ''' </summary>
    Public Class CwpProvince
        Inherits Province

        Public Property population As Long = 0
        Public Property industrialization As Long = 0
        Public Property education As Long = 0

        Public ReadOnly Property production As Long
            Get
                CalculateProduction()
                Return production_
            End Get
        End Property
        Private production_ As Long = 0

        ''' <summary>
        ''' Sets the province properties from the parsed data.
        ''' </summary>
        ''' <param name="fileName">Name of the source file of province object.</param>
        ''' <param name="parsedData">Object with the parsed data from that file.</param>
        Public Overrides Sub SetData(fileName As String, parsedData As Dictionary(Of String, Object))
            MyBase.SetData(fileName, parsedData)

            population = parsedData("population")
            industrialization = parsedData("industrialization")
            education = parsedData("education")

        End Sub

        Public Sub UpdateDaily()
            ' Change in population number, just something small for a test
            population = population * 1.00003
            ' Change in industrialization
            ' Change of education
        End Sub

        Public Sub CalculateProduction()
            production_ = population * industrialization / 100 * education / 100
        End Sub

    End Class

End Namespace

File: ColdWarGameLogic2/GameWorld/CwpProvince.cs
using System;
using System.Linq;
using OpenGSGLibrary.GameLogic;
using OpenGSGLibrary.WorldData;

namespace WorldData
{
    /// <summary>
    /// Province class for cold war game. Adds specialised properties to base province class.
    /// </summary>
    public class CwpProvince : Province
    {
        public long population { get; set; } = 0;
        public long industrialization { get; set; } = 0;
        public long education { get; set; } = 0;
        public string terrain { get; set; } = string.Empty;

        public long production
        {
            get
            {
                CalculateProduction();
                return production_;
            }
        }
        private long production_ = 0;

        public CwpProvince()
            : base() { }

        /// <summary>
        /// Sets the province properties from the parsed data.
        /// </summary>
        /// <param name="fileName">Name of the source file of province object.</param>
        /// <param name="parsedData">Object with the parsed data from that file.</param>
        public override void SetData(string fileName, Lookup<string, object> parsedData)
        {
            base.SetData(fileName, parsedData);

            population = ToLong(SingleOrDefault(parsedData, "population"));
            industrialization = ToLong(SingleOrDefault(parsedData, "industrialization"));
            education = ToLong(SingleOrDefault(parsedData, "education"));
            terrain = ToStringSafe(SingleOrDefault(parsedData, "terrain"));
        }

        /// <summary>
        /// Handler for game ticks.
        /// In this game, a tick represents a day.
        /// </summary>
        /// <param name="sender">sender TickHandler</param>
        /// <param name="e">TickEventArgs containing current game time</param>
        public override void OnTickDone(object sender, TickEventArgs e)
        {
            UpdateDaily();
        }

        private void UpdateDaily()
        {
            // Change in population number, small daily growth (preserve integer storage)
            population = Convert.ToInt64(Math.Round(population * 1.00003));
            // Change in industrialization
            // Change of education
        }

        public void CalculateProduction()
        {
            // Use floating arithmetic and convert to long like the original VB behaviour
            production_ = Convert.ToInt64(
                population * industrialization / 100.0 * education / 100.0
            );
        }

        // -- Helpers to safely extract single values from ILookup --
        private static object SingleOrDefault(ILookup<string, object> lookup, string key)
        {
            if (lookup == null)
                return null;
            return lookup.Contains(key) ? lookup[key].SingleOrDefault() : null;
        }

        private static long ToLong(object o)
        {
            if (o == null)
                return 0;
            try
            {
                return Convert.ToInt64(o);
            }
            catch
            {
                return 0;
            }
        }

        private static string ToStringSafe(object o) => o?.ToString() ?? string.Empty;
    }
}
